Composer/packages/client/src/pages/language-generation/index.tsx

```diff
@@ -35,47 +34,30 @@ interface LGPageProps extends RouteComponentProps<{}> {
 const LGPage: React.FC<LGPageProps> = props => {
   const { state } = useContext(StoreContext);
   const { lgFiles, dialogs } = state;
+
   const path = props.location?.pathname ?? '';
   const { fileId = 'common' } = props;
   const edit = /edit(\/)*$/.test(path);
   const file = lgFiles.find(({ id }) => id === 'common');
+  const navLinks = useMemo(() => {
```

Comment:
Do we need a `useMemo` here?

1. `navLinks` is computed from `dialogs`.
2. Component will be re-rendered whenever `navLinks` changes
3. Calculation inside `useMemo` will be invoked whenever `dialogs` changes

It's a traditional 'computed value' scenarios.
In a redux-style component, there will be a `mapStateToProps` handler for transforming data;
In a React FC, just call the transformer function per render.

Reply:
I don't get it, like you said "it's a typical 'computed value' scenarios" , we expect `navLinks` re-calculate only when `dialogs` changes. can you give some code change suggestion?

Reply:
> I don't get it, like you said "it's a typical 'computed value' scenarios" , we expect `navLinks` re-calculate only when `dialogs` changes. can you give some code change suggestion?

I got the difference. In your case, `dialogs` comes from StoreContext.state (didn't notice that, I thoguht it's from props, my bad).
The render process will be executed whenever store state changes, which is more frequently than we expect.

Reply:
we need the `useMemo` hook here, just share the data transformer across LG and LG page